### PR TITLE
Migrate the node project. 

### PR DESCRIPTION
**Pull request by shamil-gadelshin**
Node project was converted to the Substrate version 2.0.0rc4

Comments:
- build scripts in both runtime and node will be changed during the work on the node_spec_builder.
- tracing, telemetry, weights, benchmarking are out of the scope of the current release
- runtime API, constants, primitives were moved to separate modules
- node executor, service, RPC modules were merged in the single node crate unlike the substrate base implementation

### CODE REVIEW

File: node/src/command.rs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        2019
+    }
+
+    fn executable_name() -> &'static str {
+        "joystream-node"
+    }
+
+    fn load_spec(&self, id: &str) -> std::result::Result<Box<dyn sc_service::ChainSpec>, String> {
+        Ok(match id {
+            "dev" => Box::new(chain_spec::Alternative::Development.load().unwrap()), //TODO
```
Comment (mnaamani) [inline node/src/command.rs:58]: what is the "//TODO" about

File: runtime/src/lib.rs [hunk 2]
```diff
@@ -724,8 +609,8 @@ construct_runtime!(
         Offences: pallet_offences::{Module, Call, Storage, Event},
         RandomnessCollectiveFlip: pallet_randomness_collective_flip::{Module, Call, Storage},
         Sudo: pallet_sudo::{Module, Call, Config<T>, Storage, Event<T>},
+        Contracts: pallet_contracts::{Module, Call, Config, Storage, Event<T>},
```
Comment (mnaamani) [inline runtime/src/lib.rs:612]: Unless we are able to control who can deploy contracts, we should we keep the Contracts module out for now until we decide we need it.

File: runtime/src/runtime_api.rs [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    AccountId, AuthorityDiscoveryId, Balance, BlockNumber, EpochDuration, GrandpaAuthorityList,
+    GrandpaId, Hash, Index, RuntimeVersion, Signature, VERSION,
+};
+use crate::{
+    AllModules, AuthorityDiscovery, Babe, Call, Contracts, Grandpa, Historical, InherentDataExt,
+    RandomnessCollectiveFlip, Runtime, SessionKeys, System, TransactionPayment,
+};
+
+/// The SignedExtension to the basic transaction logic.
+pub type SignedExtra = (
```
Comment (mnaamani) [inline runtime/src/runtime_api.rs:25]: If the contracts module is in the runtime, I think there needs to be a signed extension, something about checks on Gas?

**Comment by mnaamani** [comment]
The node is building, runs and with a dev chain blocks are being produced. I'm testing with some scripts here: https://github.com/mnaamani/joystream/tree/iznik-api-examples/api-examples

Based on the latest types library and updated polkadot-js/api as per https://github.com/Joystream/joystream/pull/1124

Having issues with submitting transactions:
```
2020-08-08 17:49:46        RPC-CORE: submitExtrinsic(extrinsic: Extrinsic): Hash:: 1002: Verification Error: Execution: Could not convert parameter `tx` between node and runtime: No such variant in enum MultiSignature: RuntimeApi("Execution: Could not convert parameter `tx` between node and runtime: No such variant in
Error: 1002: Verification Error: Execution: Could not convert parameter `tx` between node and runtime: No such variant in enum MultiSignature: RuntimeApi("Execution: Could not convert parameter `tx` between node and runtime: No such variant in
```
We may not be using the correct version of polkadot-js/api or there is a misconfiguration in the node/runtime

**Comment by mnaamani** [comment]
The travis build is taking very long time to build, and is not completing, so we may need to tweak the .travis.yml file

```
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.

Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

The build has been terminated
```

**Comment by mnaamani** [comment]
> Having issues with submitting transactions:

Perhaps I'm not properly constructing types ? @Lezek123

**Comment by Lezek123** [comment]
Tried to reproduce. I used https://github.com/shamil-gadelshin/substrate-runtime-joystream/tree/migrate_runtime_module4 to start a development node and https://github.com/mnaamani/joystream/tree/iznik-api-examples/api-examples to run the scripts.

I ran:
`yarn workspace api-examples script injectDataObjects`
and then:
`yarn workspace api-examples script listDataDirectory`
But didn't run into any issues.

The output of the second one also confirmed that:
`Data Directory contains 132 objects`

**Comment by mnaamani** [comment]
> Tried to reproduce. I used https://github.com/shamil-gadelshin/substrate-runtime-joystream/tree/migrate_runtime_module4 to start a development node and https://github.com/mnaamani/joystream/tree/iznik-api-examples/api-examples to run the scripts.
> 
> I ran:
> `yarn workspace api-examples script injectDataObjects`
> and then:
> `yarn workspace api-examples script listDataDirectory`
> But didn't run into any issues.
> 
> The output of the second one also confirmed that:
> `Data Directory contains 132 objects`

Those were the steps I followed. I tried rebuilding the node a second time but I'm still getting the same issue.

File: node/src/command.rs (hunk 1 above)
Comment (shamil-gadelshin) [reply node/src/command.rs:58]: I forgot to replace the unwrap() with the "?" operator. I'll fix it.

File: runtime/src/lib.rs (hunk 2 above)
Comment (shamil-gadelshin) [reply runtime/src/lib.rs:612]: Ok. I'll remove all traces of the contracts pallet.

File: runtime/src/runtime_api.rs (hunk 3 above)
Comment (shamil-gadelshin) [reply runtime/src/runtime_api.rs:25]: I'll remove all traces of the contracts pallet.

**Comment by shamil-gadelshin** [comment]
Part of this [PR](https://github.com/Joystream/joystream/pull/1148)

**Comment by Lezek123** [comment]
I was able to reproduice @mnaamani's issue by fetching the last commit, which I'm not sure I fetched before (https://github.com/Joystream/joystream/pull/1126/commits/fb9845beb7773e2282057572488259bf32487218)

Then I re-built the runtime using https://github.com/Joystream/joystream/pull/1126/commits/10fa8dd43f556b7e253181594722fc43ef3f2964 and the issue didn't occur anymore.

That leads me to think that https://github.com/Joystream/joystream/pull/1126/commits/fb9845beb7773e2282057572488259bf32487218 introduced some changes that are causing the mentioned error.

File: runtime/src/lib.rs [hunk 4]
```diff
@@ -110,7 +106,7 @@ impl system::Trait for Runtime {
     type Hash = Hash;
     type Hashing = BlakeTwo256;
     type AccountId = AccountId;
-    type Lookup = Indices;
+    type Lookup = IdentityLookup<AccountId>;
```
Comment (mnaamani) [inline runtime/src/lib.rs:109]: With this change, as identified by @shamil-gadelshin we were required to add `"Address": "AccountId","LookupSource": "AccountId"` to the types registration in polkadot-js api, which isn't done automatically by the api library unless it detects the `node-template` specName in the runtime version.

**Comment by mnaamani** [comment]
> That leads me to think that [fb9845b](https://github.com/Joystream/joystream/commit/fb9845beb7773e2282057572488259bf32487218) introduced some changes that are causing the mentioned error.

You were correct. The change was specifically how the address lookup is done and requires a change in the types registration: https://github.com/Joystream/joystream/pull/1126/files#r469018972

**Pull request merged**